images/virtualization-artifact/pkg/controller/vm/internal/watchers/vmsnapshot_watcher.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watcher
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+type VirtualMachineSnapshotWatcher struct {
+	client client.Client
+}
+
+func NewVirtualMachineSnapshotWatcher(client client.Client) *VirtualMachineSnapshotWatcher {
+	return &VirtualMachineSnapshotWatcher{
+		client: client,
+	}
+}
+
+func (w VirtualMachineSnapshotWatcher) Watch(mgr manager.Manager, ctr controller.Controller) error {
+	return ctr.Watch(
+		source.Kind(mgr.GetCache(), &virtv2.VirtualMachineSnapshot{}),
+		handler.EnqueueRequestsFromMapFunc(w.enqueueRequests),
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return true },
+			DeleteFunc: func(e event.DeleteEvent) bool { return true },
+			UpdateFunc: w.filterUpdateEvents,
+		},
+	)
+}
+
+func (w VirtualMachineSnapshotWatcher) enqueueRequests(ctx context.Context, obj client.Object) (requests []reconcile.Request) {
+	vmSnapshot, ok := obj.(*virtv2.VirtualMachineSnapshot)
+	if !ok {
+		slog.Default().Error(fmt.Sprintf("expected a VirtualMachineSnapshot but got a %T", obj))
+		return nil
+	}
+
+	var vms virtv2.VirtualMachineList
+	// TODO use index.
+	err := w.client.List(ctx, &vms, &client.ListOptions{
+		Namespace: vmSnapshot.Namespace,
+	})
+	if err != nil {
+		slog.Default().Error(fmt.Sprintf("failed to list virtual machines: %s", err))
+		return
+	}
+
+	for _, vm := range vms.Items {
+		if vmSnapshot.Spec.VirtualMachineName == vm.Name {
+			requests = append(requests, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      vm.Name,
+					Namespace: vm.Namespace,
+				},
```

Comment:
I think this VM searching should be done in reconciler, not while enqueue requests. We can't set `Failed` on VMSnapshot resource doing VM checking this way.